lang/attribute/interface/src/lib.rs

```diff
@@ -50,11 +50,9 @@ use syn::parse_macro_input;
 ///         #[access_control(SetCount::accounts(&self, &ctx))]
 ///         pub fn set_count(&mut self, ctx: Context<SetCount>, new_count: u64) -> Result<()> {
 ///             // Ask the auth program if we should approve the transaction.
-///             let cpi_program = ctx.accounts.auth_program.clone();
-///             let cpi_ctx = CpiContext::new(cpi_program, Empty {});
-///
+///             //TODO[paulx]: what to do with this?
```

Comment:
What's the problem here? This seems fine, though I prefer moving the cpi_ctx above where it previously was.

Reply:
yea I think the todo in the `is_realized` function is the only one that actually matters. There the program comes from a remaining account instead of the Accounts struct so wasn't sure whether I can just add it as a proper field to the Account struct without breaking something because Im not super familiar yet with the `#[access_control]` and `#[interface]` attributes

also sure, I'll move cpi_ctx in back to where it was. I actually did this in a lot of places. My way of thinking about this has been that moving things into variables if they are not a) mutated or b) used at least twice is not a good idea because variables imply to the reader that either a) or b) is going to happen and it confuses the reader if neither happens (with the exception where a function call or struct becomes too confusing to read). curious what u think about that.

Reply:
So we actually need the `cpi_program` variable for `#[interface]` because the interface CPI "client" doesn't actually know what program ID to use, so it needs to be provided. See https://github.com/project-serum/anchor/blob/master/lang/attribute/interface/src/lib.rs#L216

For all other CPI's, however, this is not the case, since the program address is hardcoded in the CPI client for security reasons (facilitated with the use of `declare_id`)

So our options are

1. Keep The old API with `CpiContext::new(cpi_program, cpi_accounts)`
2. Use the new api, and make `cpi_program` an option inside `CpiContext`, e.g., `CpiContext::new(cpi_accounts).with_program(cpi_program)`.

What do you think?

Reply:
3) for interface cpi calls, add a program arg after the cpi_context and before the args expansion in the function signature

1) if we keep the old api, wouldnt all calls use cpi_program, not just interface ones? im probably misunderstanding

2) I like this too

Reply:
1) No because the cpi clients hardcode the program_id when constructing the `Instruction` struct for CPI. This is done via the `crate::ID` variable defined by `declare_id` in all Anchor programs.

Reply:
1. ah what I meant is that callers would still use cpi_program in all programs to pass it into new (even tho it is not used for the actual call)

2. seems ok to me, better break things now than later

Reply:
if we go with 2 maybe instead of an option, there could be a new type InterfaceCpiContext. for normal cpi calls u should not be able to change the program whereas for interface calls you always need to

Reply:
@fanatid thoughts?

Reply:
I do not have strong thoughts about Cpi/interface :neutral_face:

Reply:
I'd prefer to not introduce a new type, avoid the breaking change, and revisit this once we address https://github.com/project-serum/anchor/issues/454, which will determine the future of `#[interface]`.